File: src/main/java/tech/zolhungaj/amqapi/servercommands/social/PlayerProfile.java
package tech.zolhungaj.amqapi.servercommands.social;

import com.squareup.moshi.Json;
import tech.zolhungaj.amqapi.servercommands.Command;
import tech.zolhungaj.amqapi.servercommands.CommandTypeOld;
import tech.zolhungaj.amqapi.sharedobjects.AnimeList;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

public record PlayerProfile(
        @Json(name = "name")
        String nickname,
        String originalName,
        @Json(name = "guessPercent")
        GuessPercent guessPercent,
        @Json(name = "badges")
        List<DisplayedPlayerBadge> displayedBadges,
        @Json(name = "allBadges")
        List<AvailablePlayerBadge> availablePlayerBadges,
        @Json(name = "chatGlowActive")
        Optional<Boolean> chatGlowActive,
        @Json(name = "chatColorActive")
        Optional<Boolean> chatColorActive,
        @Json(name = "rankedChatGlowUnlocked")
        Optional<Boolean> rankedChatGlowUnlocked,
        @Json(name = "rankedChatColorUnlocked")
        Optional<Boolean> rankedChatColorUnlocked,
        int level,
        @Json(name = "avatarProfileImage")
        Boolean avatarProfileImage,
        @Json(name = "profileEmoteId")
        Optional<Integer> profileImageIdentifiedByEmoteId,
        @Json(name = "avatar")
        Optional<ProfileAvatar> profileImageFromAvatar,
        @Json(name = "creationDate")
        CreationDate creationDate,
        @Json(name = "list")
        DisplayedAnimeList displayedAnimeList,
        @Json(name = "songCount")
        SongCount songCount
) implements Command {

    public PlayerProfile{
        if(chatGlowActive == null){
            chatGlowActive = Optional.empty();
        }
        if(chatColorActive == null){
            chatColorActive = Optional.empty();
        }
        if(rankedChatGlowUnlocked == null){
            rankedChatGlowUnlocked = Optional.empty();
        }
        if(rankedChatColorUnlocked == null){
            rankedChatColorUnlocked = Optional.empty();
        }
    }
    @Override
    public String commandName() {
        return CommandTypeOld.PLAYER_PROFILE.commandName;
    }

    public record ProfileAvatar(
            String avatarName,
            String colorName,
            String outfitName,
            @Json(name = "optionActive")
            int optionActive,
            String optionName
    ) {
    }

    public record CreationDate(
            @Json(name = "hidden")
            boolean isHidden,
            @Json(name = "adminView")
            boolean isAdminView,
            @Json(name = "value")
            Optional<LocalDate> value
    ) {}

    public record GuessPercent(
            @Json(name = "hidden")
            boolean isHidden,
            @Json(name = "adminView")
            boolean isAdminView,
            @Json(name = "value")
            Optional<Double> value
    ) {}
    public record SongCount(
            @Json(name = "hidden")
            boolean isHidden,
            @Json(name = "adminView")
            boolean isAdminView,
            @Json(name = "value")
            Optional<Integer> value
    ) {}

    public record DisplayedAnimeList(
            @Json(name = "hidden")
            boolean isHidden,
            @Json(name = "adminView")
            boolean isAdminView,
            @Json(name = "listId")
            Optional<AnimeList> listSite,
            @Json(name = "listUserUrl")
            Optional<String> listSiteUserUrl,
            @Json(name = "listUser")
            Optional<String> listSiteUsername
    ) {
    }

    public record DisplayedPlayerBadge(
            @Json(name = "special")
            boolean isSpecial,
            @Json(name = "fileName")
            String filename,
            @Json(name = "name")
            String badgeName,
            @Json(name = "id")
            int badgeId,
            @Json(name = "slot")
            int slotId,
            @Json(name = "type")
            int badgeType,
            String unlockDescription
    ) {}

    public record AvailablePlayerBadge(
            @Json(name = "unlocked")
            Boolean isUnlocked,
            Optional<Boolean> showInChat,
            @Json(name = "slot")
            Optional<Integer> slotId,
            @Json(name = "special")
            boolean isSpecial,
            @Json(name = "fileName")
            String filename,
            @Json(name = "name")
            String badgeName,
            @Json(name = "id")
            int badgeId,
            @Json(name = "type")
            int badgeType,
            String unlockDescription
    ) {
        public AvailablePlayerBadge{
            if(isUnlocked == null){
                isUnlocked = false;
            }
            if(showInChat == null){
                showInChat = Optional.empty();
            }
            if(slotId == null){
                slotId = Optional.empty();
            }
        }
    }
}

File: src/main/java/tech/zolhungaj/amqapi/servercommands/social/PlayerProfile.kt
package tech.zolhungaj.amqapi.servercommands.social

import com.squareup.moshi.Json
import tech.zolhungaj.amqapi.servercommands.CommandType
import tech.zolhungaj.amqapi.servercommands.objects.*

@JvmRecord
@CommandType("player profile")
data class PlayerProfile(
    @Json(name = "error") val error: String?,
    @Json(name = "name") val playerName: String?,
    @Json(name = "originalName") val originalName: String?,
    @Json(name = "guessPercent") val guessPercent: ProfileGuessPercent?,
    @Json(name = "badges") val displayedBadges: List<ProfileDisplayedPlayerBadge>?,
    @Json(name = "allBadges") val availablePlayerBadges: List<ProfileAvailablePlayerBadge>?,
    @Json(name = "nameColors") val availableNameColors: List<NameDecoration>?,
    @Json(name = "nameGlows") val availableNameGlows: List<NameDecoration>?,
    @Json(name = "nameColorClass") val nameColorCssClass: String?,
    @Json(name = "nameGlowClass") val nameGlowCssClass: String?,
    @Json(name = "level") val level: Int?,
    @Json(name = "avatarProfileImage") val avatarProfileImage: Boolean?,
    @Json(name = "profileEmoteId") val profileImageIdentifiedByEmoteId: Int?,
    @Json(name = "avatar") val profileImageFromAvatar: ProfileAvatar?,
    @Json(name = "creationDate") val creationDate: ProfileCreationDate?,
    @Json(name = "list") val displayedAnimeList: ProfileDisplayedAnimeList?,
    @Json(name = "songCount") val songCount: ProfileSongCount?
)
